### ISSUE TITLE
bug for Irrigation parameter table read

### ISSUE THREAD

**Issue by cenlinhe** [NCAR/noahmp#115]
Users report this bug for WRF debug mode using Noah-MP: https://forum.mmm.ucar.edu/threads/floating-point-exception-phys-module_sf_noahmplsm-f-at-noahmp_sflx-subroutine.15638/#post-39058

Here is what from the WRF users. This needs to be fixed soon:
"
I agree, it does not make sense that the reading of namelist does not work in debug compilation mode. For sure is that, in debug mode, we can not operate with 'NaN' variables, which it is not checked in production mode and for sure compiled with optimization > -O2

I realized that the call to read the irrigation parameters (subroutine read_mp_irrigation_parameters from phys/module_sf_noahmpdrv.F) is done within NOAHMP_INIT called in phys/module_physics_init.F, which has:
Code:
if(iopt_irr  >= 1) call read_mp_irrigation_parameters()
Therfore, as in my case, if you are not using irrigation iopt_irr=0, irrigation parameters are never read. Which might explain, why it does not have any value...?
So maybe, should the NOAHMP_INIT be modified, and introduce 'default' values to the irrigation_parameters values, to overcome the situation when irrigation is not used? Something like (I use the default values as they appear in read_mp_irrigation_parameters):
Code:
    ! Default values
    IRR_FRAC_TABLE   = -1.0E36    ! irrigation Fraction
    IRR_HAR_TABLE    =  0        ! number of days before harvest date to stop irrigation
    IRR_LAI_TABLE    = -1.0E36    ! Minimum lai to trigger irrigation
    IRR_MAD_TABLE    = -1.0E36    ! management allowable deficit (0-1)
    FILOSS_TABLE     = -1.0E36    ! fraction of flood irrigation loss (0-1)
    SPRIR_RATE_TABLE = -1.0E36    ! mm/h, sprinkler irrigation rate
    MICIR_RATE_TABLE = -1.0E36    ! mm/h, micro irrigation rate
    FIRTFAC_TABLE    = -1.0E36    ! flood application rate factor
    IR_RAIN_TABLE    = -1.0E36    ! maximum precipitation to stop irrigation trigger
    if(iopt_irr  >= 1) call read_mp_irrigation_parameters()
Although it would be more elegant if irrigation_parameters is called including the iopt_irr parameter and then inside:
Code:
(...)
  IF (iopt_irr >= 1) THEN
    inquire( file='MPTABLE.TBL', exist=file_named )
    if ( file_named ) then
      open(15, file="MPTABLE.TBL", status='old', form='formatted', action='read', iostat=ierr)
(...)
    read(15,noahmp_irrigation_parameters)
    close(15)
  END IF
And modify its call from NOAHMP_INIT as:
Code:
call read_mp_irrigation_parameters(iopt_irr)
"

**Comment by LluisFB** [comment]
Dear all, 

I could find time to suggest a solution for it. I tested and it seems to be working.

In `module_sf_noahmpdrv.F`, in the `SUBROUTINE NOAHMP_INIT` I replace
```
call read_mp_optional_parameters()
    if(iopt_irr  >= 1) call read_mp_irrigation_parameters()
```
 by:
```
    call read_mp_optional_parameters()
    call read_mp_irrigation_parameters(iopt_irr)
```

Then in the module `module_sf_noahmplsm.F` I replaced:
```
subroutine read_mp_irrigation_parameters()
    implicit none
    integer :: ierr
(...)
    inquire( file='MPTABLE.TBL', exist=file_named ) 
    if ( file_named ) then
      open(15, file="MPTABLE.TBL", status='old', form='formatted', action='read', iostat=ierr)
    else
      open(15, status='old', form='formatted', action='read', iostat=ierr)
    end if
```
by:
```
  subroutine read_mp_irrigation_parameters(iopt_irr)
    implicit none

    INTEGER, INTENT(in)                                  :: iopt_irr
(....)
    IF (iopt_irr >= 1) THEN
      inquire( file='MPTABLE.TBL', exist=file_named ) 
      if ( file_named ) then
        open(15, file="MPTABLE.TBL", status='old', form='formatted', action='read', iostat=ierr)
      else
        open(15, status='old', form='formatted', action='read', iostat=ierr)
      end if
    ELSE
      ! Nothing to read, nothing to do
      RETURN

    END IF
```

**Comment by cenlinhe** [comment]
This proposed bug fix looks good to me. @CharlesZheZhang @prasanthvkrishna What do you think?

**Comment by prasanthvkrishna** [comment]
@cenlin Yes, I think the proposed bug fix looks good to me.

**Comment by CharlesZheZhang** [comment]
I agree, this proposed bug fix looks good to me.

**Comment by cenlinhe** [comment]
OK, then I will prepare a PR soon for this bug fix.

**Comment by LluisFB** [comment]
@cenlinhe I just yesterday performed a PR for another minor bug. I'm happy to do it if it helps

**Comment by cenlinhe** [comment]
@LluisFB Yes, please do it. Thank you! Please submit the PR to the "release-v4.7-WRF" branch rather than "master" or "develop" branch.

**Comment by LluisFB** [comment]
Dear @celinhe I forked from github.com/NCAR/noahmp repository, but when I open it I do not find the wrf driver.
```
 git remote -v
origin	git@github.com:LluisFB/noahmp.git (fetch)
origin	git@github.com:LluisFB/noahmp.git (push)
upstream	git@github.com:NCAR/noahmp.git (fetch)
upstream	git@github.com:NCAR/noahmp.git (push)
$ ls drivers/
hrldas  lis
```
I would appreciate if you can tell me, from which noahmp repository should I fork to proceed

**Comment by cenlinhe** [comment]
please go to branch "release-v4.7-WRF". If this is too complicated for you, feel free to let me know and I can do the PR.

**Comment by LluisFB** [comment]
@cenlinhe, sorry but I only forked the 'master' branch, I re-forked again with all the branches included

**Comment by LluisFB** [comment]
@cenlinhe I created the Pull request

**Comment by cenlinhe** [comment]
This issue has been fixed in an earlier PR commit in branch release-v4.7-WRF: https://github.com/NCAR/noahmp/commit/e5c0859874407859936739e8be8741f9aed369ee
so I am closing it now.
